drivers/ps2/ps2_npcx_controller.c

```diff
@@ -0,0 +1,354 @@
+
+#define NPCX_PS2_CH_COUNT 4
+
+/*
+ * Set WDAT3-0 and clear CLK3-0 in the PSOSIG register to
+ * reset the shift mechanism.
+ */
+#define NPCX_PS2_SHIFT_MECH_RESET (uint8_t) ~NPCX_PSOSIG_CLK_MASK_ALL
+
+/* in 50us units */
+#define PS2_RETRY_COUNT 10000
+
+#define PS2_TRANSACTION_TIMEOUT K_MSEC(20)
+
+/* Device config */
+struct ps2_npcx_ctrl_config {
+	uintptr_t base;
+	struct npcx_clk_cfg clk_cfg;
+};
+
+/* Driver data */
+struct ps2_npcx_ctrl_data {
+	/*
+	 * Bit mask to record the enabled PS/2 channels.
+	 * Only bit[7] and bit[5:3] are used
+	 * (i.e. the bit position of CLK3-0 in the PS2_PSOSIG register)
+	 */
+	uint8_t channel_enabled_mask;
+	struct k_sem lock;
+	struct k_sem tx_sync_sem;
+	ps2_callback_t callback_isr[NPCX_PS2_CH_COUNT];
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) ((const struct ps2_npcx_ctrl_config *)(dev)->config)
+
+#define DRV_DATA(dev) ((struct ps2_npcx_ctrl_data *)(dev)->data)
+
+#define HAL_PS2_INSTANCE(dev) (struct ps2_reg *)(DRV_CONFIG(dev)->base)
+
+int ps2_npcx_ctrl_configure(const struct device *dev, uint8_t channel_id,
+			    ps2_callback_t callback_isr)
+{
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
+	}
+
+	if (callback_isr == NULL) {
+		return -EINVAL;
+	}
+
+	data->callback_isr[channel_id] = callback_isr;
+	k_sem_give(&data->lock);
+
+	return 0;
+}
+
+int ps2_npcx_ctrl_enable_interface(const struct device *dev, uint8_t channel_id,
+				   bool enable)
+{
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+
+	k_sem_take(&data->lock, K_FOREVER);
+	/*
+	 * Disable the interrupt during changing the enabled channel mask to
+	 * prevent from preemption.
+	 */
+	irq_disable(DT_INST_IRQN(0));
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
+	}
+
+	if (enable) {
+		data->channel_enabled_mask |= BIT(NPCX_PSOSIG_CLK(channel_id));
+		/* Enable the relevant channel clock */
+		inst->PSOSIG |= BIT(NPCX_PSOSIG_CLK(channel_id));
+	} else {
+		data->channel_enabled_mask &= ~BIT(NPCX_PSOSIG_CLK(channel_id));
+		/* Disable the relevant channel clock */
+		inst->PSOSIG &= ~BIT(NPCX_PSOSIG_CLK(channel_id));
+	}
+
+	irq_enable(DT_INST_IRQN(0));
+	k_sem_give(&data->lock);
+
+	return 0;
+}
+
+static int ps2_npcx_ctrl_bus_busy(const struct device *dev)
+{
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+
+	/*
+	 * The driver pulls the CLK for non-active channels to low when Start
+	 * bit is detected and pull the CLK of the active channel low after
+	 * Stop bit detected. The EOT bit is set when Stop bit is detected,
+	 * but both SOT and EOT are cleared when all CLKs are pull low
+	 * (due to Shift Mechanism is reset)
+	 */
+	return (IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT) |
```

Comment:
Should that be || ?

Alternatively, can you use 

IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT | NPCX_PSTAT_EOT) ?

Reply:
Done by changing from "|" to "||".